bench/wtperf/wtperf_throttle.c

```diff
@@ -0,0 +1,108 @@
+/*-
+ * Public Domain 2014-2015 MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "wtperf.h"
+
+/*
+ * Put the initial config together for running a throttled workload.
+ */
+int
+setup_throttle(CONFIG_THREAD *thread) {
+	THROTTLE_CONFIG *throttle_cfg;
+	throttle_cfg = &thread->throttle_cfg;
+
+	/* If the interval is very small, we do laps of 1 */
```

Comment:
I'd done some experimenting along these lines initially. The problem I found was that the timing becomes less accurate when we are doing very few operations per run through the loop. 

For example, if I cut down and merge the two bottom clauses I get the following output when I attempt to do 1000 reads per second over 2 threads and 247600 inserts per second over 3 threads.
```
9294 reads, 3368119 inserts, 0 updates, 34 truncates, 0 checkpoints in 5 secs (5 total secs)
9317 reads, 3371651 inserts, 0 updates, 34 truncates, 0 checkpoints in 5 secs (10 total secs)
9330 reads, 3377288 inserts, 0 updates, 33 truncates, 0 checkpoints in 5 secs (15 total secs)
```

Whereas with the three clauses I get the following with the same configuration as above.
```
10000 reads, 3677246 inserts, 0 updates, 37 truncates, 0 checkpoints in 5 secs (5 total secs)
10000 reads, 3661618 inserts, 0 updates, 37 truncates, 0 checkpoints in 5 secs (10 total secs)
10000 reads, 3650866 inserts, 0 updates, 36 truncates, 0 checkpoints in 5 secs (15 total secs)
10000 reads, 3649161 inserts, 0 updates, 37 truncates, 0 checkpoints in 5 secs (20 total secs)
10000 reads, 3592797 inserts, 0 updates, 36 truncates, 0 checkpoints in 5 secs (25 total secs)
```
I had initially not had the initial very small values clause at the top, but found that there was also another hiccough at these very low values, so I added a third clause.